lib/pf/access_filter/radius.pm

```diff
@@ -0,0 +1,166 @@
+
+use Log::Log4perl;
+use pf::api::jsonrpcclient;
+use pf::violation qw (violation_view_top);
+use pf::locationlog qw(locationlog_set_session);
+
+use base qw(pf::access_filter);
+tie our %ConfigRadiusFilters, 'pfconfig::cached_hash', 'config::RadiusFilters';
+tie our %RadiusFilterEngineScopes, 'pfconfig::cached_hash', 'FilterEngine::RadiusScopes';
+
+
+=head1 SUBROUTINES
+
+=over
+
+=item new
+
+=cut
+
+sub new {
+   my $logger = Log::Log4perl::get_logger("pf::access_filter::radius");
+   $logger->debug("instantiating new pf::access_filter::radius");
+   my ( $class, %argv ) = @_;
+   my $self = bless {}, $class;
+   return $self;
+}
+
+=item test
+
+Test all the rules
+
+=cut
+
+sub test {
+    my ($self, $scope, $args) = @_;
+    if (exists $RadiusFilterEngineScopes{$scope}) {
+       $args->{'violation'} = violation_view_top($args->{'mac'});
+       return $RadiusFilterEngineScopes{$scope}->match_first($args);
+    }
+    return undef;
+}
+
+=head2 handleRoleInRule
+
+Handles the role update
+
+=cut
+
+sub handleAnswerInRule {
+    my ($self, $rule, $args) = @_;
+    my $logger = Log::Log4perl::get_logger( ref($self) );
+    my $radius_reply_ref = {};
+    if(defined $rule) {
+        $radius_reply_ref = {'Reply-Message' => "Inverse is the best"};
+        my $i = 1;
+        $logger->info(evalParam($rule->{'log'},$args)) if defined($rule->{'log'});
+        while (1) {
+            if (defined($rule->{"answer$i"}) && $rule->{"answer$i"} ne '') {
+                my @answer = split('=>',$rule->{"answer$i"});
+                $args->{'session_id'} = setSession($args) if ($answer[1] =~ /\$session_id/);
+                evalAnswer(\@answer,$args,\$radius_reply_ref);
+            } else {
+                last;
+            }
+            $i++;
+        }
+    }
+    return ($radius_reply_ref);
+}
+
+sub setSession {
+    my($args) = @_;
+    my (%session_id);
+    pf::web::util::session(\%session_id,undef,6);
+    $session_id{client_mac} = $args->{'mac'};
+    $session_id{wlan} = $args->{'ssid'};
+    $session_id{switch_id} = $args->{'switch'}{'_id'};
+    locationlog_set_session($args->{'mac'}, $session_id{_session_id});
+    return ($session_id{_session_id});
+}
+
+=head2 evalAnswer
+
+evaluate the radius answer
+
+=cut
+
+sub evalAnswer {
+    my ($answer,$args,$radius_reply_ref) = @_;
+
+    my $return = evalParam(@{$answer}[1],$args);
+    my @multi_value = split(';',$return);
+    @{$answer}[0] =~ s/\s//g;
+    if (scalar @multi_value > 1) {
+        $$radius_reply_ref->{@{$answer}[0]} = \@multi_value;
+    } else {
+        $$radius_reply_ref->{@{$answer}[0]} = $return;
+    }
+
+}
+    
+=head2 evalParam
+
+evaluate all the variables
+
+=cut
+
+sub evalParam {
+    my ($answer, $args) = @_;
+    $answer =~ s/\$([a-zA-Z_]+)/$args->{$1} \/\/ ''/ge;
+    $answer =~ s/\${([a-zA-Z_\-]+)\.([a-zA-Z_\-]+)\.([a-zA-Z_\-]+)}/$args->{$1}->{$2}{$3} \/\/ ''/ge;
+    $answer =~ s/\${([a-zA-Z_\-]+)\.([a-zA-Z_\-]+)}/$args->{$1}->{$2} \/\/ ''/ge;
```

Comment:
Isn't there a way to make this more dynamic...

What if I want to access the 4th level... We'll have to copy/paste again